policies/recipes/steps.go

```diff
@@ -90,8 +100,335 @@ func parsePermissions(s string) (os.FileMode, error) {
+		_, err = os.Stat(filedir)
+		if os.IsNotExist(err) {
+			err = os.MkdirAll(filedir, 0755)
+		}
+		if err != nil {
+			return err
+		}
+		fmt.Printf("os.Create %s\n", filen)
+		reader, err := f.Open()
+		if err != nil {
+			return err
+		}
+
+		dst, err := os.OpenFile(filen, os.O_RDWR|os.O_CREATE, 0755)
+		if err == nil {
+			_, err = io.Copy(dst, reader)
+		}
+		reader.Close()
+
+		if err != nil {
+			return err
+		}
+		err = os.Chtimes(filen, time.Now(), f.Modified)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func extractTar(tarName string, dst string) error {
+	file, err := os.Open(tarName)
+	if err != nil {
+		return err
+	}
+	tr := tar.NewReader(file)
+
+	// Check for conflicts
+	for {
+		header, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		filen, err := common.NormPath(filepath.Join(dst, header.Name))
+		if err != nil {
+			return err
+		}
+		fmt.Printf("checking if file %s exists\n", filen)
+		stat, err := os.Stat(filen)
+		if os.IsNotExist(err) {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+		if header.Typeflag == tar.TypeDir && stat.IsDir() {
+			// it's ok if directories already exist
+			continue
+		}
+		return fmt.Errorf("file exists: %s", filen)
+	}
+
+	file.Seek(0, 0)
+	tr = tar.NewReader(file)
+
+	// Create dirs
+	for {
+		var err error
+		header, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		if header.Typeflag != tar.TypeDir {
+			continue
+		}
+		filen, err := common.NormPath(filepath.Join(dst, header.Name))
+		if err != nil {
+			return err
+		}
+		_, err = os.Stat(filen)
+		if err == nil {
+			continue
+		}
+		if !os.IsNotExist(err) {
+			return err
+		}
+		err = os.MkdirAll(filen, os.FileMode(header.Mode))
+		if err != nil {
+			return err
+		}
+		err = os.Chown(filen, header.Uid, header.Gid)
+		if err != nil {
+			return err
+		}
+	}
+
+	file.Seek(0, 0)
+	tr = tar.NewReader(file)
+
+	// Create files.
+	for {
+		var err error
+		header, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		filen, err := common.NormPath(filepath.Join(dst, header.Name))
+		if err != nil {
+			return err
+		}
+		filedir := filepath.Dir(filen)
+		_, err = os.Stat(filedir)
+		if err != nil {
+			err = os.MkdirAll(filedir, 0755)
+		}
+		if err != nil {
+			return err
+		}
+		switch header.Typeflag {
+		case tar.TypeDir:
+			continue
+		case tar.TypeReg, tar.TypeRegA:
+			fmt.Printf("os.Create %s (owner %s/%d group %s/%d)\n", filen, header.Uname, header.Uid, header.Gname, header.Gid)
+			dst, err := os.Create(filen)
+			if err == nil {
+				_, err = io.Copy(dst, tr)
+			}
+		case tar.TypeLink:
+			err = os.Link(header.Linkname, filen)
+			continue
+		case tar.TypeSymlink:
+			err = os.Symlink(header.Linkname, filen)
+			continue
+		case tar.TypeChar:
+			err = unix.Mknod(filen, uint32(unix.S_IFCHR), int(unix.Mkdev(uint32(header.Devmajor), uint32(header.Devminor))))
```

Comment:
this too was just where my working code was at the time-it seems particularly difficult to reimplement `mknod` in go, we should do our due diligence and look for others who have solved this problem. i'm not very confident in this being the 'right' way to do it

Reply:
mknod seems to map fairly directly to a syscall so calling that (either through the unix or syscall packages) seems to be correct.

Reply:
it's more about the way the arguments are constructed. i recall having a hard time getting that to work and not finding good go docs on it.

Reply:
Yeah, I'm in the same boat. As far as I can tell this works but I don't have enough expertise in unix block/char devices to know. Is there anyone else in particular who would know this area?